### PR TITLE
Decouple UT From Server Executors

### PR DESCRIPTION
**Pull request by Maithem**
## Overview
The infrastructure unit tests are written in a way that's tightly
coupled to the server's executors. This makes it hard to change
the servers threading model without changing all the corresponding
tests. This patch breaks this coupling.

Why should this be merged: removes some unit test hacks and it is a prerequisite for #2264

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/clients/ClientMsgHandler.java [hunk 1]
```diff
@@ -77,8 +77,8 @@ public boolean handle(CorfuMsg message, ChannelHandlerContext ctx) {
                 if (ret != null) {
                     client.getRouter().completeRequest(message.getRequestID(), ret);
                 }
-            } catch (Exception ex) {
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/clients/ClientMsgHandler.java]: Document why we need an error here (some reflection code throws Errors)

File: test/src/test/java/org/corfudb/infrastructure/BaseServerTest.java [hunk 2]
```diff
@@ -1,12 +1,15 @@
...
-import org.assertj.core.api.Assertions;
 import org.corfudb.protocols.wireprotocol.CorfuMsg;
 import org.corfudb.protocols.wireprotocol.CorfuMsgType;
 import org.junit.Test;
 
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.RejectedExecutionException;
 
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
```
Comment (annym) [inline test/src/test/java/org/corfudb/infrastructure/BaseServerTest.java]: Codacy Issue found: [Unused import - org.assertj.core.api.Assertions.assertThatThrownBy.](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651439)

**Comment by annym** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 4
           

Complexity decreasing per file
==============================
+ test/src/test/java/org/corfudb/infrastructure/TestServerRouter.java  -3
         

Clones added
============
- test/src/test/java/org/corfudb/infrastructure/SequencerServerTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651439&bid=15577748)

File: test/src/test/java/org/corfudb/infrastructure/LayoutServerTest.java [hunk 3]
```diff
@@ -14,7 +16,13 @@
 import org.corfudb.protocols.wireprotocol.LayoutCommittedRequest;
 import org.corfudb.protocols.wireprotocol.LayoutMsg;
 import org.corfudb.protocols.wireprotocol.LayoutPrepareRequest;
+import org.corfudb.protocols.wireprotocol.LayoutPrepareResponse;
 import org.corfudb.protocols.wireprotocol.LayoutProposeRequest;
+import org.corfudb.protocols.wireprotocol.LayoutProposeResponse;
```
Comment (annym) [inline test/src/test/java/org/corfudb/infrastructure/LayoutServerTest.java]: Codacy Issue found: [Unused import - org.corfudb.protocols.wireprotocol.LayoutProposeResponse.](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651439)

Comment (annym) [inline test/src/test/java/org/corfudb/infrastructure/LayoutServerTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.corfudb.protocols.wireprotocol.LayoutProposeResponse'](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651439)

File: test/src/test/java/org/corfudb/infrastructure/BaseServerTest.java (hunk 2 above)
Comment (annym) [inline test/src/test/java/org/corfudb/infrastructure/BaseServerTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.assertj.core.api.Assertions.assertThatThrownBy'](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651439)

**Comment by annym** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 4
           

Complexity decreasing per file
==============================
+ test/src/test/java/org/corfudb/infrastructure/TestServerRouter.java  -3
         

Clones added
============
- test/src/test/java/org/corfudb/infrastructure/SequencerServerTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651439&bid=15577748)

File: test/src/test/java/org/corfudb/infrastructure/LayoutServerTest.java (hunk 3 above)
Comment (annym) [inline test/src/test/java/org/corfudb/infrastructure/LayoutServerTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.corfudb.protocols.wireprotocol.LayoutProposeResponse'](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651439)

Comment (annym) [inline test/src/test/java/org/corfudb/infrastructure/LayoutServerTest.java]: Codacy Issue found: [Unused import - org.corfudb.protocols.wireprotocol.LayoutProposeResponse.](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651439)

File: test/src/test/java/org/corfudb/infrastructure/BaseServerTest.java (hunk 2 above)
Comment (annym) [inline test/src/test/java/org/corfudb/infrastructure/BaseServerTest.java]: Codacy Issue found: [Unused import - org.assertj.core.api.Assertions.assertThatThrownBy.](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651439)

Comment (annym) [inline test/src/test/java/org/corfudb/infrastructure/BaseServerTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.assertj.core.api.Assertions.assertThatThrownBy'](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651439)

**Comment by annym** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
           

Complexity decreasing per file
==============================
+ test/src/test/java/org/corfudb/infrastructure/TestServerRouter.java  -3
         

Clones added
============
- test/src/test/java/org/corfudb/infrastructure/SequencerServerTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651439&bid=15577748)

**Comment by annym** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
           

Complexity decreasing per file
==============================
+ test/src/test/java/org/corfudb/infrastructure/TestServerRouter.java  -3
         

Clones added
============
- test/src/test/java/org/corfudb/infrastructure/SequencerServerTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/CorfuDB/CorfuDB/pullRequest?prid=4653412&bid=15584299)

File: test/src/test/java/org/corfudb/runtime/view/LayoutViewTest.java [hunk 4]
```diff
@@ -276,6 +276,7 @@ public void getConsensusFromCurrentMembers1Node() throws Exception {
         corfuRuntime.getLayoutView().getRuntimeLayout(l).sealMinServerSet();
         corfuRuntime.getLayoutView().updateLayout(newLayout, 1L);
 
+        System.out.println("check12");
```
Comment (medhavidhawan) [inline test/src/test/java/org/corfudb/runtime/view/LayoutViewTest.java]: is this intentional ?

File: test/src/test/java/org/corfudb/runtime/view/LayoutSealTest.java [hunk 5]
```diff
@@ -124,13 +141,12 @@ public void successfulChainSeal() {
...
-            e.printStackTrace();
-        }
-        assertLayoutEpochs(2, 2, 2);
-        assertServerRouterEpochs(2, 2, 2, 2, 2);
+        runtimeLayout.sealMinServerSet();
+
+        int[] ports = l.getAllServers().stream().mapToInt(s -> Integer.parseInt(s.split(":")[1])).toArray();
+        Map<Long, Integer> epochs = collectServerEpochs(ports);
+        // Check that we have a quorum of "sealed epochs"
+        assertThat(epochs.get(l.getEpoch())).isGreaterThanOrEqualTo((ports.length / 2) + 1);
```
Comment (medhavidhawan) [inline test/src/test/java/org/corfudb/runtime/view/LayoutSealTest.java]: is this correct. These things are fluent but not very expressive so asking

File: test/src/test/java/org/corfudb/infrastructure/LayoutServerTest.java [hunk 6]
```diff
@@ -256,8 +262,7 @@ public void checkPaxosPhasesPersisted() throws Exception {
         assertThat(s2).isPhase1Rank(new Rank(1L, AbstractServerTest.testClientId));
 
         // validate phase2 data persistence
-        sendPropose(newEpoch, 1, newLayout);
-        Assertions.assertThat(getLastMessage().getMsgType()).isEqualTo(CorfuMsgType.ACK);
+        Assertions.assertThat(sendPropose(newEpoch, 1, newLayout).join()).isTrue();
```
[the remark this answers is not in the document]
Comment (Maithem) [reply test/src/test/java/org/corfudb/infrastructure/LayoutServerTest.java]: Yes. True means that it was ACK'd.

File: runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java [hunk 7]
```diff
@@ -558,6 +558,8 @@ public void prefixTrim(final Token address) {
                                 + "Invalidate layout for this client and retry, attempt: {}/{}",
                         address, runtimeEpoch, serverEpoch, x + 1, numRetries);
                 runtime.invalidateLayout();
+            }  catch (CompletionException e) {
```
[the remark this answers is not in the document]
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java]: Removed.

File: test/src/test/java/org/corfudb/infrastructure/AbstractServerTest.java [hunk 8]
```diff
@@ -47,16 +52,40 @@ public void setServer(AbstractServer server) {
...
+
+    public <T> CompletableFuture<T> sendRequest(CorfuMsg msg) {
+        msg.setClientID(testClientId)
+                .setRequestID(requestCounter.getAndIncrement())
+                .setEpoch(0L);
+        clientRouter.setClientID(testClientId);
+        return clientRouter.sendMessageAndGetCompletable(msg);
+    }
+
+    public TestClientRouter getClientRouter() {
```
Comment (medhavidhawan) [inline test/src/test/java/org/corfudb/infrastructure/AbstractServerTest.java]: Should this create a new instance everytime ?

File: test/src/test/java/org/corfudb/infrastructure/LayoutServerTest.java [hunk 9]
```diff
@@ -40,8 +46,10 @@ public LayoutServer getDefaultServer() {
      */
     @Test
     public void nonBootstrappedServerNoLayout() {
-        requestLayout(0);
-        Assertions.assertThat(getLastMessage().getMsgType()).isEqualTo(CorfuMsgType.LAYOUT_NOBOOTSTRAP);
+        //requestLayout(0);
```
Comment (medhavidhawan) [inline test/src/test/java/org/corfudb/infrastructure/LayoutServerTest.java]: please remove commented code if not needed

File: test/src/test/java/org/corfudb/infrastructure/LayoutServerTest.java [hunk 10]
```diff
@@ -190,15 +197,15 @@ public void checkLayoutPersisted() throws Exception {
...
-        setEpoch(NEW_EPOCH);
+        setEpoch(NEW_EPOCH).join();
 
         // Start the process of electing a new layout. But that layout will not take effect
         // till it is committed.
-        sendPrepare(NEW_EPOCH, 1);
-        Assertions.assertThat(getLastMessage().getMsgType()).isEqualTo(CorfuMsgType.LAYOUT_PREPARE_ACK);
+        LayoutPrepareResponse resp = sendPrepare(NEW_EPOCH, 1).join();
+        Assertions.assertThat(resp.getLayout()).isNull();
+        Assertions.assertThat(resp.getRank()).isEqualTo(-1);
```
Comment (medhavidhawan) [inline test/src/test/java/org/corfudb/infrastructure/LayoutServerTest.java]: The -1 should be a named constant. It has a significance. This should be also done in the actual code.

**Pull request merged**